addons/crm/wizard/base_partner_merge.py

```diff
@@ -282,9 +289,14 @@ def _merge(self, partner_ids, dst_partner=None):
         _logger.info("dst_partner: %s", dst_partner.id)
 
         # FIXME: is it still required to make and exception for account.move.line since accounting v9.0 ?
-        if SUPERUSER_ID != self.env.uid and 'account.move.line' in self.env and self.env['account.move.line'].sudo().search([('partner_id', 'in', [partner.id for partner in src_partners])]):
+        if check_move_lines and 'account.move.line' in self.env and self.env['account.move.line'].sudo().search([('partner_id', 'in', [partner.id for partner in src_partners])]):
             raise UserError(_("Only the destination contact may be linked to existing Journal Items. Please ask the Administrator if you need to merge several contacts linked to existing Journal Items."))
 
+        # Update the company of origin partners to the dest one if needed
+        if force_company:
+            for partner in partner_ids:
+                partner.company_id = dst_partner.company_id
```

Comment:
You could `partner_ids.write({'company_id': dst_partner.company_id.id})` instead of looping :)

Reply:
Note about this part: it may not be enough as some overwrite of `write` on the partner will kick in... (cfr https://github.com/odoo/odoo/blob/3ab25b60bc5e13a91a518f5effd46a28b026a8f7/odoo/addons/base/res/res_partner.py#L500-L506)

If this is to work consistently, it has to make sure that
1/ The dst company is allowed for all possibly linked users
2/ All users linked to partner in the merging set should have their current company changed to the dst company